compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java

```diff
@@ -0,0 +1,1138 @@
+
+        if (bLangNode instanceof BLangSimpleVarRef) {
+            if (((BLangSimpleVarRef) bLangNode).symbol != null) {
+                return getTypeFromBType(((BLangSimpleVarRef) bLangNode).symbol.getType());
+            }
+        }
+
+        if (bLangNode instanceof BLangUserDefinedType) {
+            if (((BLangUserDefinedType) bLangNode).symbol != null) {
+                return getTypeFromBType(((BLangUserDefinedType) bLangNode).symbol.getType());
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ExplicitAnonymousFunctionExpressionNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (bLangNode instanceof BLangFunction) {
+            return Optional.of(typesFactory.getTypeDescriptor(((BLangFunction) nodeFinder.lookup
+                    (this.bLangCompilationUnit, node.lineRange())).returnTypeNode.getBType()));
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ImplicitAnonymousFunctionExpressionNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (!(bLangNode instanceof BLangArrowFunction)) {
+            return Optional.empty();
+        }
+
+        BType returnType = ((BLangArrowFunction) nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange()))
+                .funcType.getReturnType();
+        if (!node.rightDoubleArrow().isMissing() &&
+                node.rightDoubleArrow().lineRange().endLine().offset() <= linePosition.offset()) {
+            // Cursor is at the expression node
+            return Optional.of(typesFactory.getTypeDescriptor(returnType));
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(PositionalArgumentNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (bLangNode instanceof BLangRecordLiteral) {
+            return getExpectedType(bLangNode);
+        }
+
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(FunctionDefinitionNode node) {
+        Optional<ReturnTypeDescriptorNode> returnTypeDesc = node.functionSignature().returnTypeDesc();
+        if (returnTypeDesc.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<Symbol> functionSymbol = this.semanticModel.symbol(node);
+
+        if (functionSymbol.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return ((FunctionSymbol) functionSymbol.get()).typeDescriptor().returnTypeDescriptor();
```

Comment:
This is an unconfirmed cast. Shall we check the kind of functionSymbol in line 451 to ensure that CCE won't happen